spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -16,117 +16,119 @@
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
-
+import com.google.api.gax.rpc.NotFoundException;
 import com.google.cloud.secretmanager.v1beta1.AccessSecretVersionResponse;
-import com.google.cloud.secretmanager.v1beta1.ProjectName;
-import com.google.cloud.secretmanager.v1beta1.Secret;
 import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
-import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient.ListSecretsPagedResponse;
 import com.google.cloud.secretmanager.v1beta1.SecretVersionName;
 import com.google.protobuf.ByteString;
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
 
 import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
 import org.springframework.core.env.EnumerablePropertySource;
 
 /**
- * Retrieves secrets from GCP Secret Manager under the current GCP project.
+ * A property source for Secret Manager which accesses the Secret Manager APIs when {@link #getProperty} is called.
  *
  * @author Daniel Zou
  * @author Eddú Meléndez
  * @since 1.2.2
  */
 public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
 
-	private static final Log LOGGER = LogFactory.getLog(SecretManagerPropertySource.class);
-
-	private static final String LATEST_VERSION_STRING = "latest";
-
-	private final Map<String, Object> properties;
+	private static final String GCP_SECRET_PREFIX = "gcp-secret/";
 
-	private final String[] propertyNames;
+	private final GcpProjectIdProvider projectIdProvider;
 
 	public SecretManagerPropertySource(
 			String propertySourceName,
 			SecretManagerServiceClient client,
-			GcpProjectIdProvider projectIdProvider,
-			String secretsPrefix) {
-		this(propertySourceName, client, projectIdProvider, secretsPrefix, Collections.EMPTY_MAP);
-	}
-
-	public SecretManagerPropertySource(
-			String propertySourceName,
-			SecretManagerServiceClient client,
-			GcpProjectIdProvider projectIdProvider,
-			String secretsPrefix,
-			Map<String, String> versions) {
+			GcpProjectIdProvider projectIdProvider) {
 		super(propertySourceName, client);
 
-		Map<String, Object> propertiesMap = createSecretsPropertiesMap(
-				client, projectIdProvider.getProjectId(), secretsPrefix, versions);
+		this.projectIdProvider = projectIdProvider;
+	}
 
-		this.properties = propertiesMap;
-		this.propertyNames = propertiesMap.keySet().toArray(new String[propertiesMap.size()]);
+	@Override
+	public Object getProperty(String name) {
+		SecretVersionName secretIdentifier = parseFromProperty(name, this.projectIdProvider);
+
+		if (secretIdentifier != null) {
+			return getSecretPayload(secretIdentifier);
+		}
+		else {
+			return null;
+		}
 	}
 
+	/**
+	 * The {@link SecretManagerPropertySource} is not enumerable, so this always returns an empty array.
+	 * @return the empty array.
+	 */
 	@Override
 	public String[] getPropertyNames() {
-		return propertyNames;
+		return new String[0];
```

Comment:
Interesting that this still works.
Maybe we should log a warning when this method is called?

Reply:
Can't unfortunately. This method gets called by framework code somewhere and will pollute the logs.

Reply:
Maybe at `DEBUG` or `TRACE` level? How often will this be called?

Reply:
Hmm I don't think it's worth it; feel like it will just add noise. It is called at least 10 times per test run in the integration tests.